### PR TITLE
[test] Print the driver logs at runtime

### PR DESCRIPTION
**Pull request by SongGuyang**
## Why are these changes needed?

Currently when the driver(triggered by `run_string_as_driver`) hangs, we can't see the logs in the screen. In this PR, we start a thread to print the driver logs at runtime.

### CODE REVIEW

**Comment by SongGuyang** [comment]
Who can review this small PR?

**Comment by rkooo567** [comment]
Hmm I don't quite understand the problem. Why does it hang in the first place?

**Comment by SongGuyang** [comment]
> Hmm I don't quite understand the problem. Why does it hang in the first place?

`run_string_as_driver` will wait for the driver process to exit. In some cases(maybe caused by some bugs), the driver will not exit and the test will time out. It is hard to diagnose because we can't see the driver log in buildkite screen.

File: python/ray/_private/test_utils.py [hunk 1]
```diff
@@ -220,14 +221,28 @@ def run_string_as_driver(driver_script: str, env: Dict = None, encode: str = "ut
         stderr=subprocess.STDOUT,
         env=env,
     )
+    out = ""
+
+    def catch_output(stdout):
+        for line in iter(stdout.readline, b""):
```
Comment (rkooo567) [inline python/ray/_private/test_utils.py:227]: Is `stdout.readline` threadsafe? Also, is it possible this for loop terminates (reaches EoF) when stdout is written slowly? E.g., there's no more stdout, for loop terminates, and then new output is written to stdout

File: python/ray/_private/test_utils.py [hunk 2]
```diff
@@ -220,14 +221,28 @@ def run_string_as_driver(driver_script: str, env: Dict = None, encode: str = "ut
...
+            decode_line = ray._private.utils.decode(line, encode_type=encode)
+            nonlocal out
+            out += decode_line
+            print("[driver]", decode_line)
```
Comment (rkooo567) [inline python/ray/_private/test_utils.py:231]:
```suggestion
            print("[driver] ", decode_line)
```

File: python/ray/_private/test_utils.py [hunk 3]
```diff
@@ -220,14 +221,28 @@ def run_string_as_driver(driver_script: str, env: Dict = None, encode: str = "ut
...
+
+    # Start a thread to print the driver logs at runtime.
+    t = Thread(target=catch_output, args=(proc.stdout,))
+    t.daemon = True
```
Comment (rkooo567) [inline python/ray/_private/test_utils.py:235]: Doesn't need to be a daemon since we join anyway?

**Pull request closed without merging**